The meshes of icub3 seems not compatible with rviz2

When trying to visualize the icub3 meshes on `rviz2` i got these errors:

```
[ERROR] [1670425894.974567837] [rviz2]: The STL file 'package://iCub/meshes/simmechanics/sim_icub3_torso_2_prt.stl' is malformed. It starts with the word 'solid' and also contains the word 'endsolid', indicating that it's an ASCII STL file, but rviz can only load binary STL files so it will not be loaded. Please convert it to a binary STL file.
```

`ros2 pkg list | grep iCub` works fine.

cc @pattacini @traversaro

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,13 +10,13 @@ on:
         required: true
         default: 'false'
       branch:
-        description: 'Branch of icub-models to which the action will push'     
+        description: 'Branch of icub-models to which the action will push'
         required: true
         default: 'devel'
-        
+
   schedule:
   # * is a special character in YAML so you have to quote this string
-  # Execute a "nightly" build at 2 AM UTC 
+  # Execute a "nightly" build at 2 AM UTC
   - cron:  '0 2 * * *'
 
 env:
@@ -35,13 +35,13 @@ jobs:
 
     steps:
     - uses: actions/checkout@v2
-    
+
     - name: Change ICUB_MODELS_BRANCH if necessary
       shell: bash
       if: ${{ github.event_name == 'workflow_dispatch' }}
       run: |
         echo "ICUB_MODELS_BRANCH=${{ github.event.inputs.branch }}" >> $GITHUB_ENV
-   
+
     # Print environment variables to simplify development and debugging
     - name: Environment Variables
       shell: bash
@@ -61,7 +61,7 @@ jobs:
         cd ${GITHUB_WORKSPACE}
         sudo apt-get install  --assume-yes --force-yes python-lxml python-yaml python-numpy python-setuptools
         # probably python on the path return a python interpreter and the find_package(PythonInterp) in CMake another,
-        # let's install both debian packages and pip packages to be sure 
+        # let's install both debian packages and pip packages to be sure
         sudo pip install lxml numpy pyyaml catkin_pkg
         # install urdf_parser_py and save the last commit SHA1 hash
         git clone $URDF_PARSER_PY_REPOSITORY_URL
@@ -117,7 +117,7 @@ jobs:
         # Clone icub-models repo
         git clone -b $ICUB_MODELS_BRANCH https://github.com/${DEPLOYMENT_REPOSITORY}.git ${ICUB_MODELS_SOURCE_DIR}
         cmake --build . --target copy-models-to-icub-models
-        # Generate commit message 
+        # Generate commit message
         echo "Automatic build. GitHub Actions build: $GITHUB_RUN_ID" >> ${GITHUB_WORKSPACE}/deploy_commit_message
         echo "icub-model-generator commit: robotology/icub-models-generator@$GITHUB_SHA" >> ${GITHUB_WORKSPACE}/deploy_commit_message
         echo "urdf_parser_py commit: ros/urdf_parser_py@$URDF_PARSER_PY_COMMIT" >> ${GITHUB_WORKSPACE}/deploy_commit_message

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -54,7 +54,7 @@ jobs:
     - name: Generate models
       run: |
         sudo apt-get update
-        sudo apt-get install libeigen3-dev libace-dev libtinyxml-dev libxml2-dev
+        sudo apt-get install ruby libeigen3-dev libace-dev libtinyxml-dev libxml2-dev
         # Save the url of the repository and the user-name of the committ author
         export CURRENT_REPOSITORY_URL=`git remote get-url origin`
         # Start in the parent directory of icub-model-generator

--- a/simmechanics/CMakeLists.txt
+++ b/simmechanics/CMakeLists.txt
@@ -444,3 +444,10 @@ add_custom_command(TARGET generate-models-simmechanics
                    COMMAND ${CMAKE_COMMAND} -E copy_directory "${CMAKE_CURRENT_SOURCE_DIR}/data/components/left_hand_mk3/conf" "${CMAKE_BINARY_DIR}/${BUILD_PREFIX}/conf_left_hand_mk3"
                    COMMAND ${CMAKE_COMMAND} -E copy           "${CMAKE_CURRENT_SOURCE_DIR}/data/components/left_hand_mk3/model.config" "${CMAKE_BINARY_DIR}/${BUILD_PREFIX}/robots/left_hand_mk3/."
                    COMMENT "Copying Simmechanics meshes")
+
+# Convert meshes from ASCII to binary
+add_custom_command(TARGET generate-models-simmechanics
+                   POST_BUILD
+                   WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                   COMMAND ruby ./scripts/convertSTL.rb ${CMAKE_BINARY_DIR}/${BUILD_PREFIX}/meshes/simmechanics/*.stl
+                   COMMENT "Converting Simmechanics meshes to binary")

--- a/simmechanics/scripts/README.md
+++ b/simmechanics/scripts/README.md
@@ -1,7 +1,8 @@
-Scripts to generate the 3D positions of iCub skin taxels from the skin triangles centers extracted from CAD models. 
+Scripts to generate the 3D positions of iCub skin taxels from the skin triangles centers extracted from CAD models.
 
-## Dependencies 
+## Dependencies
+- [Ruby](https://www.ruby-lang.org/en/)
 * [`YARP` Python bindings](http://www.yarp.it/yarp_swig.html)
 * [`iDynTree` Python bindings](https://github.com/robotology/idyntree#installation)
-* `numpy` 
+* `numpy`
 * `scipy`

--- a/simmechanics/scripts/convertSTL.rb
+++ b/simmechanics/scripts/convertSTL.rb
@@ -1,0 +1,96 @@
+#!/usr/bin/env ruby
+# convertSTL.rb - Converts STL files between binary and ASCII encoding
+# by Chris Polis
+#
+# This script detects which encoding the stl file is in and converts it to
+#  the opposite encoding and saves the file as *-ascii.stl or *-binary.stl.
+#  I wrote this script to save disk space and bandwidth when using stl files.
+#
+# USAGE:
+#  $ ruby convertSTL.rb [filename(s) of .stl to be converted]
+#  or 'chmod +x' and run as ./convertSTL.rb
+#
+#  note: path wildcards, (e.g.: "*.stl") are supported and all matching
+#        files will be converted
+
+
+# Helper methods
+class Float
+    def to_sn # to scientific notation
+      "%E" % self
+    end
+
+    def self.from_sn str # generate a float from scientific notation
+      ("%f" % str).to_f
+    end
+  end
+
+  # Pass in filename as only argument
+  if ARGV.size == 0
+    puts "Usage: ./converSTL.rb [stl filename]"
+    exit
+  end
+
+  begin
+    # Read all files matching arg, "foo.stl", "*.stl", "../stl/*", etc...
+    ARGV.each do |originalFilename|
+      original = File.new(originalFilename, "r")
+
+      # Read first line - check binary or ASCII
+      tempLine = original.gets
+      if tempLine.start_with? "solid"
+        outFilename = originalFilename.sub(/\.stl/i, '-binary.stl')
+        puts "#{originalFilename} is in ASCII format, converting to BINARY"
+        outFile = File.new(outFilename, "w")
+        outFile.write("\0" * 80) # 80 bit header - ignored
+        outFile.write("FFFF")   # 4 bit integer # of triangles - filled later
+        triCount = 0
+
+        # ASCII STL format (from Wikipedia):
+        # solid name(optional)
+        #
+        # [foreach triangle]
+        # facet normal ni nj nk
+        # outer loop
+        # vertex v1x v1y v1z
+        # vertex v2x v2y v2z
+        # vertex v3x v3y v3z
+        # endloop
+        # endfacet
+        # endsolid name(optional)
+
+        while temp = original.gets
+          next if temp =~ /^\s*$/ or temp.include? 'endsolid' # ignore whitespace
+          temp.sub! /facet normal/, ''
+          normal = temp.split(' ').map{ |num| Float.from_sn num }
+          triCount += 1
+          temp = original.gets # 'outer loop'
+
+          temp = original.gets
+          vertexA = temp.sub(/vertex/, '').split(' ').map{ |num| Float.from_sn num }
+          temp = original.gets
+          vertexB = temp.sub(/vertex/, '').split(' ').map{ |num| Float.from_sn num }
+          temp = original.gets
+          vertexC = temp.sub(/vertex/, '').split(' ').map{ |num| Float.from_sn num }
+
+          temp = original.gets # 'endsolid'
+          temp = original.gets # 'endfacet'
+
+          outFile.write(normal.pack("FFF"))
+          outFile.write(vertexA.pack("FFF"))
+          outFile.write(vertexB.pack("FFF"))
+          outFile.write(vertexC.pack("FFF"))
+          outFile.write("\0\0")
+        end
+        outFile.seek(80, IO::SEEK_SET)
+        outFile.write([ triCount ].pack("V"))
+        outFile.close
+        File.rename(outFilename, originalFilename)
+      else
+        puts "#{originalFilename} is in BINARY format, skipping"
+      end
+      original.close
+    end
+  rescue => error
+    puts "Error: #{error}"
+  end

--- a/simmechanics/scripts/convertSTL.rb
+++ b/simmechanics/scripts/convertSTL.rb
@@ -1,4 +1,15 @@
 #!/usr/bin/env ruby
+
+# Copyright 2021 Chris Polis
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
 # convertSTL.rb - Converts STL files between binary and ASCII encoding
 # by Chris Polis
 #

